src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -676,12 +709,29 @@ int RGWRemoteDataLog::read_log_info(const DoutPrefixProvider *dpp, rgw_datalog_i
 int RGWRemoteDataLog::read_source_log_shards_info(const DoutPrefixProvider *dpp, map<int, RGWDataChangesLogInfo> *shards_info)
 {
   rgw_datalog_info log_info;
+
   int ret = read_log_info(dpp, &log_info);
   if (ret < 0) {
     return ret;
   }
 
-  return run(dpp, new RGWReadRemoteDataLogInfoCR(&sc, log_info.num_shards, shards_info));
+  set<int> shards;
+  auto iter = shards.begin();
+  for (int i = 0; i < (int)log_info.num_shards; i++)
+    iter = shards.insert(iter, i);
+
+  static constexpr int MAX_RETRIES = 10;
+  for (int i = 0; i < MAX_RETRIES; i++) {
+    ret = run(dpp, new RGWReadRemoteDataLogInfoCR(&sc, &shards, shards_info));
```

Comment:
Yup agree that it's the better place to handle it.
This original commit was to limit the changes only to the radosgw-admin command part so I wouldn't inadvertently introduce inappropriate logic to other workflows, until some experienced rgw developers suggested otherwise, which I got here. Thanks! 
I'll make the changes.

Reply:
> wouldn't inadvertently introduce inappropriate logic to other workflows

i see what you mean. for the sync logic, i do think it's much better to handle these retries locally instead of relying on sync's higher-level retry mechanisms. those tend to be a lot more expensive, potentially requiring writes to the data sync error repo, and re-reads of sync status and replication logs

Reply:
Changes have been committed to move the retry logic down to `RGWReadRemoteDataLogShardInfoCR` and `RGWReadRemoteMDLogShardInfoCR`. And also add the retry to `RGWListRemoteDataLogShardCR` and `RGWListRemoteMDLogShardCR` as well. Ready to be reviewed again.